discretisedfield/field.py

```diff
@@ -2358,46 +2308,54 @@ def diff(self, direction, order=1, restrict2valid=True):
         if order not in (1, 2):
             raise NotImplementedError(f"Derivative of {order=} is not implemented.")
 
-        # Boundary conditions
-        if direction in self.mesh.bc:  # PBC
-            bc = "pbc"
-        elif self.mesh.bc == "":
-            bc = None
+        direction_idx = self.mesh.region._dim2index(direction)
+
+        # If periodic is True, we pad the field
+        if periodic:
+            field = self.pad({direction: (1, 1)}, mode="wrap")
         else:
-            bc = self.mesh.bc
+            field = self
 
-        direction_idx = self.mesh.region._dim2index(direction)
-        out = np.zeros_like(self.array)
+        valid = field.valid if restrict2valid else np.ones_like(field.valid, dtype=bool)
+
+        out = np.zeros_like(field.array)
         # Use only valid values for the derivative if restrict2valid is True
         # or use all values if restrict2valid is False
-        valid = self.valid if restrict2valid else np.ones_like(self.valid, dtype=bool)
 
         # Loop over all cells in the plane perpendicular to the direction of the
         # derivative. Use sel method in a way that keeps the dimensionality but
         # only returns a single layer in the direction of the derivative
-        point = self.mesh.region.pmin[direction_idx] + self.mesh.cell[direction_idx] / 2
-        for idx in self.mesh.sel(**{direction: (point, point)}).indices:
+        point = (
+            field.mesh.region.pmin[direction_idx] + field.mesh.cell[direction_idx] / 2
+        )
+        for idx in field.mesh.sel(**{direction: (point, point)}).indices:
             idx = list(idx)
             idx[direction_idx] = slice(None)
             valid_arr = valid[tuple(idx)]
             # Loop over all value dimensions of the vector field and
             # compute the derivative for each value dimension.
-            for dim in range(self.nvdim):
+            for dim in range(field.nvdim):
                 out[tuple([*idx, dim])] = _split_diff_combine(
-                    self.array[tuple([*idx, dim])],
+                    field.array[tuple([*idx, dim])],
                     valid_arr,
                     order,
-                    self.mesh.cell[direction_idx],
-                    bc,
+                    field.mesh.cell[direction_idx],
                 )
 
+        # Remove the padding if periodic is True
+        if periodic:
+            slices = dfu.assemble_index(
+                slice(None), field.mesh.region.ndim + 1, {direction_idx: slice(1, -1)}
+            )
```

Comment:
I think we could also use `self.mesh.region2slices`. Might be a little more obvious. (Please test first!)